### PR TITLE
Add zones implementation to the cloud controller

### PR DESCRIPTION
**Pull request by cezarsa**

### CODE REVIEW

File: cloudstack/instances.go [hunk 1]
```diff
@@ -251,7 +251,7 @@ func (cs *CSCloud) getNodeByProviderID(id string) (*node, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to list nodes: %v", err)
 	}
-	if len(kubeNodes.Items) > 0 {
```
Comment (andrestc) [inline cloudstack/instances.go:254]: :O

File: cloudstack/zones.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// GetZone returns the Zone containing the current failure zone and locality region that the program is running in
+// In most cases, this method is called from the kubelet querying a local metadata service to aquire its zone.
+// For the case of external cloud providers, use GetZoneByProviderID or GetZoneByNodeName since GetZone
+// can no longer be called from the kubelets.
+func (cs *CSCloud) GetZone() (cloudprovider.Zone, error) {
+	return cloudprovider.Zone{}, nil
+}
+
+// GetZoneByProviderID returns the Zone containing the current zone and locality region of the node specified by providerId
+// This method is particularly used in the context of external cloud providers where node initialization must be down
```
Comment (andrestc) [inline cloudstack/zones.go]: down => done
kubelets => kubelet

File: cloudstack/instances.go (hunk 1 above)
Comment (cezarsa) [reply cloudstack/instances.go:254]: I only noticed this because it exploded in the tests.

File: cloudstack/zones.go (hunk 2 above)
Comment (cezarsa) [reply cloudstack/zones.go]: Fixing, this was a copy/paste from the interface definition in kubernetes.

**Review by andrestc — APPROVED** [review]

File: cloudstack/zones.go (hunk 2 above)
Comment (cezarsa) [reply cloudstack/zones.go]: Kept kubelets as it means all kubelets in all nodes.

**Pull request merged**